### DISCUSSION TITLE
Is there no way to easily import the types generated by Prisma?

### DISCUSSION THREAD

**Discussion by andreasqlight** [prisma/prisma#12453] [Q&A] [answered]
I'd like to import the types created and use them in say, the constructor of a class or a simple method. I'm using the `PromiseReturnType` to get the type. However, this is going to add `| null` to the type. Which is fine, but in some cases there is absolutely no way the method is called without the object does not exist, so a check has been done previously. Maybe that is bad practice but kinda bloated. 

It's also a bit awkward when you want to build up interfaces that includes these types. I can see that the types are defined in index.d.ts. 

Current type delcaration:
`export type ControllerFetchType = Prisma.PromiseReturnType<typeof fetchController>;`

**Comment by austincrim** [comment, marked as answer]
Hey @andreasqlight 👋 ,

You should be able to import any of Prisma's generated types from the `@prisma/client` package.

```prisma
model User {
 ...
}
```

```ts
import type { User } from '@prisma/client'
```

Let me know if that works!

**Reply by Download** [reply]
Is there a way to get all types, without hardcoding import statements for each of them?

Try this:

```ts
import type * as Prisma from '@prisma/client';
// Org is a model in this example
type Test = Prisma.Org; // works
type Test2 = Prisma['Org']; // Cannot use namespace 'Prisma' as a type. ts(2709)
type Test3 = keyof typeof Prisma; // "Prisma" | "PrismaClient"
```

I can get at the names of the models via this code:

```ts
import { PrismaClient } from "@prisma/client";
type ModelNames = keyof {
  [P in keyof PrismaClient as P extends `$${string}` ? never : Capitalize<P>]: null
}
// type ModelNames = "Org" | "User" | "Note" | ...
```

However, I found no way to get at the actual types, even after finding their names.

Could you guys maybe export a mapped type for us that we can use? Or am I missing some way to do it?

**Reply by kanzitelli** [reply]
You might also consider using the approach from the [prisma-models](https://github.com/kanzitelli/prisma-models) package, which returns a single source type with all your Prisma Models and Enums:

```tsx
import {Prisma, $Enums} from '@prisma/client';
import {PrismaModels, PrismaEnums} from 'prisma-models';

export type Models = PrismaModels<Prisma.ModelName, Prisma.TypeMap>;
export type Enums  = PrismaEnums <typeof $Enums>;

// Now you can use it like Models['User'] and Enums['UserStatus']
```

**Comment by mcunningham** [comment]
Similar to @Download, I was looking to get a mapped typescript type of all Prisma models without importing each one individually. For anyone also looking to do this, here is what I came up with:

```typescript
import type { Prisma, PrismaClient } from "@prisma/client";

type ModelNames = Prisma.ModelName; // "User" | "Post"

export type PrismaModels = {
  [M in ModelNames]: Exclude<
    Awaited<ReturnType<PrismaClient[Uncapitalize<M>]["findUnique"]>>,
    null
  >;
};

type UserId = PrismaModels["User"]["id"] // string
```

`ModelNames` is a union of the Model names directly from Prisma's generated types.

Once I have the union of names then I create a mapped type over all of the Model Names and map it to the `Awaited` `ReturnType` of the `findUnique` call for each corresponding Model key of the `PrismaClient`. This will in turn give back the correct model types of an individual model. Since PrismaClient is expecting a lowercase value for the key/property ie PrismaClient.user I use `Uncapitalize<M>` so its properly referenced.

It would be nice to have a Mapped type directly from the generated types,  but this has been working for my use case for now.

**Comment by gearonixx** [comment]
nice!

**Comment by roger-tbg** [comment]
How does this work/interoperate with: https://www.prisma.io/docs/orm/prisma-client/type-safety#type-utilities

I can't seem to reconcile both.

**Comment by Download** [comment]
@roger-tbg I am hoping it would allow us to replace the hardcoded `prisma.post` from this example:

```ts
type PostCreateBody = Prisma.Args<typeof prisma.post, 'create'>['data']
```

To something like this:

```ts
type PostCreateBody = Prisma.Args<PrismaModels["Post"], 'create'>['data']
```

**Comment by AmodeusR** [comment]
> You should be able to import any of Prisma's generated types from the @prisma/client package.

Indeed it's possible, but keys that shouldn't be modified or asked for, like ID or createdAt are imported as required. How can I import from Prisma without these valued which are already filled by Prisma itself?

**Comment by mrdexign** [comment]
Hi!

This is a better solution:

```typescript
type ModelNames = Capitalize<Exclude<Extract<keyof PrismaClient, string>, `$${string}`>>;
```

This type includes only the model names and excludes any `symbol` types.

**Comment by PwntasticKev** [comment]
I was not able to get @austincrim 's sleek answer. What worked for me was making a file:

_types.ts_
`import {truck_stop as TruckStop} from '@prisma/client';
export type {TruckStop};
`
On my frontend code I'm importing the type. 
`import {TruckStop} from '@/types/api/types';`

**Comment by jsgv** [comment]
Using [`Exclude`](https://www.typescriptlang.org/docs/handbook/utility-types.html#excludeuniontype-excludedmembers) was exactly what I needed.

```typescript
type Props = {
    document: Exclude<Prisma.PromiseReturnType<typeof getDocument>, null>;
};
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
